File: solutions/cpp/Count_inversions.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

int CountSubarrayInversions(int, int, vector<int>*);
int MergeSortAndCountInversionsAcrossSubarrays(int, int, int, vector<int>*);

// @include
int CountInversions(vector<int> A) {
  return CountSubarrayInversions(0, A.size(), &A);
}

// Return the number of inversions in (*A_ptr)[start : end - 1].
int CountSubarrayInversions(int start, int end, vector<int>* A_ptr) {
  if (end - start <= 1) {
    return 0;
  }

  int mid = start + ((end - start) / 2);
  return CountSubarrayInversions(start, mid, A_ptr) +
         CountSubarrayInversions(mid, end, A_ptr) +
         MergeSortAndCountInversionsAcrossSubarrays(start, mid, end, A_ptr);
}

// Merge two sorted subarrays (*A_ptr)[start : mid - 1] and (*A_ptr)[mid : end
// - 1] into (*A_ptr)[start : end - 1] and return the number of inversions
// across (*A_ptr)[start : mid - 1] and (*A_ptr)[mid : end - 1].
int MergeSortAndCountInversionsAcrossSubarrays(int start, int mid, int end,
                                               vector<int>* A_ptr) {
  vector<int> sorted_A;
  int left_start = start, right_start = mid, inversion_count = 0;

  vector<int>& A = *A_ptr;
  while (left_start < mid && right_start < end) {
    if (A[left_start] <= A[right_start]) {
      sorted_A.emplace_back(A[left_start++]);
    } else {
      // A[left_start : mid - 1] are the inversions of A[right_start].
      inversion_count += mid - left_start;
      sorted_A.emplace_back(A[right_start++]);
    }
  }
  copy(A.begin() + left_start, A.begin() + mid, back_inserter(sorted_A));
  copy(A.begin() + right_start, A.begin() + end, back_inserter(sorted_A));

  // Updates A with sorted_A.
  copy(sorted_A.begin(), sorted_A.end(), A.begin() + start);
  return inversion_count;
}
// @exclude

// O(n^2) check of inversions.
template <typename T>
int N2Check(const vector<T>& A) {
  int count = 0;
  for (size_t i = 0; i < A.size(); ++i) {
    for (size_t j = i + 1; j < A.size(); ++j) {
      if (A[i] > A[j]) {
        ++count;
      }
    }
  }
  cout << count << endl;
  return count;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    vector<int> A;
    for (size_t i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(-1000000, 1000000);
      A.emplace_back(dis(gen));
    }
    assert(N2Check(A) == CountInversions(A));
  }
  return 0;
}

File: solutions/python/count_inversions.py
import sys
import random


# @include
def count_inversions(A):
    # Return the number of inversions in A[start:end].
    def count_subarray_inversions(start, end):
        # Merge two sorted subarrays A[start:mid] and A[mid:end] into
        # A[start:end] and return the number of inversions across A[start:mid]
        # and A[mid:end].
        def merge_sort_and_count_inversions_across_subarrays(start, mid, end):
            sorted_A = []
            left_start, right_start, inversion_count = start, mid, 0

            while left_start < mid and right_start < end:
                if A[left_start] <= A[right_start]:
                    sorted_A.append(A[left_start])
                    left_start += 1
                else:
                    # A[left_start:mid] are the inversions of A[right_start].
                    inversion_count += mid - left_start
                    sorted_A.append(A[right_start])
                    right_start += 1

            # Updates A with sorted_A.
            A[start:end] = sorted_A + A[left_start:mid] + A[right_start:end]
            return inversion_count

        if end - start <= 1:
            return 0
        mid = start + ((end - start) // 2)
        return (
            count_subarray_inversions(start, mid) +
            count_subarray_inversions(mid, end) +
            merge_sort_and_count_inversions_across_subarrays(start, mid, end))

    return count_subarray_inversions(0, len(A))
# @exclude


# O(n^2) check of inversions.
def n2_check(A):
    count = sum(A[i] > A[j] for i in range(len(A))
                for j in range(i + 1, len(A)))
    print(count)
    return count


def main():
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10000)
        A = [random.randint(-1000000, 1000000) for _ in range(n)]
        assert n2_check(A) == count_inversions(A)


if __name__ == '__main__':
    main()
